drivers/ethernet/phy/phy_qualcomm_ar8031.c

```diff
@@ -0,0 +1,522 @@
+{
+	const struct qc_ar8031_config *config = dev->config;
+	int ret;
+
+	ret = mdio_write(config->mdio_dev, config->addr, reg_addr, (uint16_t)data);
+	if (ret) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int qc_ar8031_mmd_set_device(const struct device *dev, uint32_t device, uint32_t addr,
+				    uint16_t mode)
+{
+	uint16_t reg_value = (device & MII_MMD_ACR_DEVAD_MASK) | MII_MMD_ACR_ADDR;
+
+	if (qc_ar8031_write(dev, MII_MMD_ACR, reg_value) < 0) {
+		return -EIO;
+	}
+	if (qc_ar8031_write(dev, MII_MMD_AADR, addr) < 0) {
+		return -EIO;
+	}
+	/* Set Function mode of data access(b01~11) and device address. */
+	if (qc_ar8031_write(dev, MII_MMD_ACR, (device & MII_MMD_ACR_DEVAD_MASK) | mode) < 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int qc_ar8031_mmd_read(const struct device *dev, uint32_t device, uint32_t addr,
+			      uint32_t *data)
+{
+	int ret;
+
+	*data = 0U;
+
+	ret = qc_ar8031_mmd_set_device(dev, device, addr, MII_MMD_ACR_DATA_NO_POS_INC);
+	if (ret) {
+		return -EIO;
+	}
+
+	ret = qc_ar8031_read(dev, MII_MMD_AADR, data);
+
+	return ret;
+}
+
+static int qc_ar8031_mmd_write(const struct device *dev, uint32_t device, uint32_t addr,
+			       uint32_t data)
+{
+	int ret;
+
+	ret = qc_ar8031_mmd_set_device(dev, device, addr, MII_MMD_ACR_DATA_NO_POS_INC);
+	if (ret) {
+		return -EIO;
+	}
+
+	ret = qc_ar8031_write(dev, MII_MMD_AADR, data);
+
+	return ret;
+}
+
+static int qc_ar8031_update_link_state(const struct device *dev)
+{
+	const struct qc_ar8031_config *const cfg = dev->config;
+	struct qc_ar8031_data *const data = dev->data;
+	bool link_up;
+	uint32_t reg_value;
+	uint16_t speed, duplex;
+
+	if (qc_ar8031_read(dev, PHY_SPECIFIC_STATUS_REG, &reg_value) < 0) {
+		return -EIO;
+	}
+
+	link_up = (uint16_t)reg_value & SPEC_STATUS_REG_LINK_MASK;
+
+	/* If there is no change in link state don't proceed. */
+	if (link_up == data->state.is_up) {
+		return -EAGAIN;
+	}
+
+	data->state.is_up = link_up;
+
+	/* If link is down, there is nothing more to be done */
+	if (data->state.is_up == false) {
+		return 0;
+	}
+
+	if (qc_ar8031_read(dev, PHY_SPECIFIC_STATUS_REG, &reg_value) < 0) {
+		return -EIO;
+	}
+
+	speed = reg_value & SPEC_STATUS_REG_SPEED_MASK;
+	duplex = reg_value & SPEC_STATUS_REG_DUPLEX_MASK;
+
+	switch (speed | duplex) {
+	case PHY_SPEED_10M | PHY_DUPLEX_FULL:
+		data->state.speed = LINK_FULL_10BASE_T;
+		break;
+	case PHY_SPEED_10M | PHY_DUPLEX_HALF:
+		data->state.speed = LINK_HALF_10BASE_T;
+		break;
+	case PHY_SPEED_100M | PHY_DUPLEX_FULL:
+		data->state.speed = LINK_FULL_100BASE_T;
+		break;
+	case PHY_SPEED_100M | PHY_DUPLEX_HALF:
+		data->state.speed = LINK_HALF_100BASE_T;
+		break;
+	case PHY_SPEED_1000M | PHY_DUPLEX_FULL:
+		data->state.speed = LINK_FULL_1000BASE_T;
+		break;
+	case PHY_SPEED_1000M | PHY_DUPLEX_HALF:
+		data->state.speed = LINK_HALF_1000BASE_T;
+		break;
+	}
+
+	LOG_INF("PHY (%d) Link speed %s Mb, %s duplex", cfg->addr,
```

Comment:
I would think `LOG_DBG` is desired here.

Reply:
Updated, thanks.